File: app/local_plugins/agent_ast_translate/android/src/main/kotlin/com/aiagent/agent_ast_translate/AstTranslateAgentSession.kt
package com.aiagent.agent_ast_translate

import android.content.Context
import android.util.Log
import com.aiagent.local_db.AppDatabase
import com.aiagent.local_db.entity.MessageEntity
import com.aiagent.plugin_interface.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow

/**
 * AstTranslateAgentSession — AST (端到端语音翻译) Agent 原生实现
 *
 * 使用 NativeAstService 通过 WebSocket / WebRTC 进行实时语音翻译。
 * 服务端完成 ASR → 翻译 → TTS 全流程，客户端只负责音频收发。
 *
 * 简化状态机：IDLE → CONNECTED → IDLE（与 StsChatAgentSession 相同模式）
 *
 * 桥接策略（AST 五件套 → STT/LLM 事件）：
 *   - recognizing(SOURCE)     → SttEventData(partialResult)
 *   - recognized(SOURCE)      → SttEventData(finalResult, requestId) + DB.user
 *   - recognizing(TRANSLATED) → LlmEventData(firstToken, textDelta, requestId)
 *   - recognized(TRANSLATED)  → LlmEventData(firstToken, textDelta, requestId)
 *   - recognitionEnd          → LlmEventData(done, fullText) + DB.assistant
 */
class AstTranslateAgentSession : NativeAgent {

    companion object {
        private const val TAG = "AstTranslateSession"
    }

    override val agentType = "ast"

    private enum class State { IDLE, CONNECTED, ERROR }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val _state = MutableStateFlow(State.IDLE)

    private lateinit var astService: NativeAstService
    private lateinit var db: AppDatabase
    private lateinit var eventSink: AgentEventSink
    private lateinit var config: NativeAgentConfig
    private lateinit var appContext: Context

    private var connectJob: Job? = null
    private var inputMode: String = "text"

    /** 当前轮次最后一段 translated 定稿文本 — recognitionEnd 时写入 done.fullText */
    @Volatile private var lastTranslatedText: String = ""
    @Volatile private var activeTranslationRequestId: String? = null

    // ─────────────────────────────────────────────────
    // NativeAgent 接口实现
    // ─────────────────────────────────────────────────

    override fun initialize(config: NativeAgentConfig, eventSink: AgentEventSink, context: Context) {
        this.config = config
        this.eventSink = eventSink
        this.inputMode = config.inputMode
        this.appContext = context.applicationContext
        this.db = AppDatabase.getInstance(context)

        astService = NativeServiceRegistry.createAst(config.astVendor ?: "volcengine")
        astService.initialize(config.astConfigJson ?: "{}", context)

        Log.d(TAG, "initialized: agentId=${config.agentId} astVendor=${config.astVendor}")
    }

    override fun connectService() {
        Log.d(TAG, "connectService: agentId=${config.agentId}")
        connectJob?.cancel()

        astService = NativeServiceRegistry.createAst(config.astVendor ?: "volcengine")
        astService.initialize(config.astConfigJson ?: "{}", appContext)

        connectJob = scope.launch {
            try {
                astService.connect(astCallback)
            } catch (e: CancellationException) {
                // normal cancellation
            } catch (e: Exception) {
                Log.e(TAG, "AST connect failed: ${e.message}")
                transitionTo(State.ERROR)
                eventSink.onError(config.agentId, "ast_connect_error", e.message ?: "Unknown error", null)
                eventSink.onAgentReady(
                    config.agentId, ready = false,
                    errorCode = "ast_connect_error",
                    errorMessage = e.message ?: "Unknown error",
                )
            }
        }
    }

    override fun disconnectService() {
        Log.d(TAG, "disconnectService: agentId=${config.agentId}")
        connectJob?.cancel()
        astService.release()
        transitionTo(State.IDLE)
        eventSink.onConnectionStateChanged(config.agentId, "disconnected")
    }

    override fun sendText(requestId: String, text: String) {
        Log.w(TAG, "sendText called but AST is voice-only, ignoring: $text")
    }

    override fun startListening() {
        Log.d(TAG, "startListening: delegating to startAudio")
        astService.startAudio()
    }

    override fun stopListening() {
        Log.d(TAG, "stopListening: delegating to stopAudio")
        astService.stopAudio()
    }

    override fun setInputMode(mode: String) {
        Log.d(TAG, "setInputMode: $mode")
        inputMode = mode
        when (mode) {
            // call 模式 = external audio：PCM 由 CallTranslationSession 通过
            // startExternalAudio + pushExternalAudioFrame 喂进来，**绝不能**在这里
            // 调 startAudio()，否则会进入 self-mic 模式与 external audio 互斥，
            // 后续 startExternalAudio 抛 "external audio cannot mix with self-mic"。
            "call" -> { /* no-op: 等 CallTranslationSession 调 startExternalAudio */ }
            "short_voice" -> { /* 按住说话，由 UI 调用 startListening/stopListening */ }
            else -> astService.stopAudio()
        }
    }

    override fun interrupt() {
        astService.interrupt()
        transitionTo(State.CONNECTED)
    }

    override fun release() {
        connectJob?.cancel()
        scope.cancel()
        astService.release()
    }

    // ─────────────────────────────────────────────────
    // 外部音频源（通话翻译等场景）—— 透传到底层 NativeAstService
    // ─────────────────────────────────────────────────

    override fun externalAudioCapability(): ExternalAudioCapability =
        astService.externalAudioCapability()

    override fun startExternalAudio(format: ExternalAudioFormat, sink: ExternalAudioSink) {
        astService.startExternalAudio(format, sink)
    }

    override fun pushExternalAudioFrame(frame: ByteArray) {
        astService.pushExternalAudioFrame(frame)
    }

    override fun stopExternalAudio() {
        astService.stopExternalAudio()
    }

    // ─────────────────────────────────────────────────
    // AST 回调 → AgentEventSink 事件转换
    // ─────────────────────────────────────────────────

    private val astCallback = object : AstCallback {
        override fun onConnected() {
            transitionTo(State.CONNECTED)
            eventSink.onConnectionStateChanged(config.agentId, "connected")
            eventSink.onAgentReady(config.agentId, ready = true)
            Log.d(TAG, "AST connected, inputMode=$inputMode")
            // call 模式下 CallTranslationSession 会随后调 startExternalAudio 把
            // 耳机解码出来的 PCM 喂进来，这里不要触发 self-mic 路径。
        }

        override fun onDisconnected() {
            transitionTo(State.IDLE)
            eventSink.onConnectionStateChanged(config.agentId, "disconnected")
            Log.d(TAG, "AST disconnected")
        }

        override fun onRecognitionStart(role: AstRole, requestId: String) {
            // No-op — chat provider lazily creates the message on the first
            // partial / firstToken event.
        }

        override fun onRecognizing(role: AstRole, requestId: String, text: String) {
            when (role) {
                AstRole.SOURCE -> {
                    eventSink.onSttEvent(SttEventData(
                        config.agentId,
                        requestId = "",
                        kind = "partialResult",
                        text = text,
                    ))
                }
                AstRole.TRANSLATED -> {
                    activeTranslationRequestId = requestId
                    lastTranslatedText = text
                    eventSink.onLlmEvent(LlmEventData(
                        config.agentId,
                        requestId = requestId,
                        kind = "firstToken",
                        textDelta = text,
                    ))
                }
            }
        }

        override fun onRecognized(role: AstRole, requestId: String, text: String) {
            when (role) {
                AstRole.SOURCE -> {
                    eventSink.onSttEvent(SttEventData(
                        config.agentId,
                        requestId = requestId,
                        kind = "finalResult",
                        text = text,
                    ))
                    persistMessage(requestId, role = "user", content = text)
                }
                AstRole.TRANSLATED -> {
                    activeTranslationRequestId = requestId
                    lastTranslatedText = text
                    eventSink.onLlmEvent(LlmEventData(
                        config.agentId,
                        requestId = requestId,
                        kind = "firstToken",
                        textDelta = text,
                    ))
                }
            }
        }

        override fun onRecognitionDone(role: AstRole, requestId: String) {
            // No-op — round-level closure is signalled by recognitionEnd.
        }

        override fun onRecognitionEnd(requestId: String) {
            val transId = activeTranslationRequestId
            val transText = lastTranslatedText
            if (transId != null && transText.isNotBlank()) {
                eventSink.onLlmEvent(LlmEventData(
                    config.agentId,
                    requestId = transId,
                    kind = "done",
                    fullText = transText,
                ))
                persistMessage(transId, role = "assistant", content = transText)
            }
            activeTranslationRequestId = null
            lastTranslatedText = ""
        }

        override fun onRecognitionError(requestId: String?, role: AstRole?, code: String, message: String) {
            eventSink.onError(config.agentId, code, message, requestId)
        }

        override fun onError(code: String, message: String) {
            transitionTo(State.ERROR)
            eventSink.onConnectionStateChanged(config.agentId, "error", message)
            eventSink.onError(config.agentId, code, message, null)
            eventSink.onAgentReady(config.agentId, ready = false, errorCode = code, errorMessage = message)
        }
    }

    // ─────────────────────────────────────────────────
    // 内部方法
    // ─────────────────────────────────────────────────

    private fun persistMessage(id: String, role: String, content: String) {
        scope.launch {
            runCatching {
                val now = System.currentTimeMillis()
                db.messageDao().insert(MessageEntity(
                    id = id,
                    agentId = config.agentId,
                    role = role,
                    content = content,
                    status = "done",
                    createdAt = now,
                    updatedAt = now,
                ))
            }
        }
    }

    private fun transitionTo(newState: State) {
        _state.value = newState
        eventSink.onStateChanged(
            config.agentId,
            newState.name.lowercase(),
            null,
        )
    }
}

File: app/local_plugins/agents/agent_ast_translate/ios/Classes/AstTranslateAgentSession.swift
import Foundation
import ai_plugin_interface
import os.log

/// AST (end-to-end speech translation) agent session — iOS port of the Kotlin
/// `AstTranslateAgentSession`. Persistence stays on the Dart side; this class
/// only bridges `AstCallback` five-tuple events to `AgentEventSink` STT/LLM
/// events (same mapping as Android).
///
/// Bridging rules:
///   - recognizing(.source)     → SttEventData(partialResult)
///   - recognized(.source)      → SttEventData(finalResult, requestId)
///   - recognizing(.translated) → LlmEventData(firstToken, textDelta, requestId)
///   - recognized(.translated)  → LlmEventData(firstToken, textDelta, requestId)
///   - recognitionEnd           → LlmEventData(done, fullText)
public final class AstTranslateAgentSession: NativeAgent {

    public let agentType = "ast-translate"

    private static let log = OSLog(subsystem: "com.aiagent.agent_ast_translate", category: "Session")

    fileprivate enum State: String { case idle, connected, error }

    private let stateLock = NSLock()
    private var state: State = .idle

    private var config: NativeAgentConfig!
    private weak var eventSink: AgentEventSink?
    private var inputMode: String = "text"

    private var astService: NativeAstService?

    /// Last translated-track snapshot for the current round. Emitted as
    /// `done.fullText` on `recognitionEnd`.
    private var lastTranslatedText: String = ""
    private var activeTranslationRequestId: String?

    public init() {}

    // ── NativeAgent lifecycle ─────────────────────────────────────────────

    public func initialize(config: NativeAgentConfig, eventSink: AgentEventSink) {
        self.config = config
        self.eventSink = eventSink
        self.inputMode = config.inputMode

        do {
            let svc = try NativeServiceRegistry.shared.createAst(config.astVendor ?? "volcengine")
            svc.initialize(configJson: config.astConfigJson ?? "{}")
            self.astService = svc
            os_log("initialized agentId=%{public}@ astVendor=%{public}@ inputMode=%{public}@",
                   log: Self.log, type: .debug,
                   config.agentId, config.astVendor ?? "volcengine", config.inputMode)
        } catch {
            os_log("createAst failed: %{public}@",
                   log: Self.log, type: .error, error.localizedDescription)
            eventSink.onError(
                sessionId: config.agentId,
                errorCode: "ast_init_failed",
                message: error.localizedDescription,
                requestId: nil
            )
        }
    }

    public func connectService() {
        os_log("connectService: agentId=%{public}@", log: Self.log, type: .debug, config.agentId)

        do {
            let svc = try NativeServiceRegistry.shared.createAst(config.astVendor ?? "volcengine")
            svc.initialize(configJson: config.astConfigJson ?? "{}")
            self.astService = svc
        } catch {
            os_log("connectService createAst failed: %{public}@",
                   log: Self.log, type: .error, error.localizedDescription)
            transitionTo(.error)
            eventSink?.onError(
                sessionId: config.agentId,
                errorCode: "ast_init_failed",
                message: error.localizedDescription,
                requestId: nil
            )
            eventSink?.onAgentReady(
                sessionId: config.agentId,
                ready: false,
                errorCode: "ast_init_failed",
                errorMessage: error.localizedDescription
            )
            return
        }

        astService?.connect(callback: AstCallbackAdapter(owner: self))
    }

    public func disconnectService() {
        os_log("disconnectService: agentId=%{public}@", log: Self.log, type: .debug, config.agentId)
        astService?.release()
        transitionTo(.idle)
        eventSink?.onConnectionStateChanged(
            sessionId: config.agentId,
            state: "disconnected",
            errorMessage: nil
        )
    }

    public func sendText(requestId: String, text: String) {
        os_log("sendText ignored — AST is voice-only", log: Self.log, type: .info)
    }

    public func startListening() {
        astService?.startAudio()
    }

    public func stopListening() {
        astService?.stopAudio()
    }

    public func setInputMode(_ mode: String) {
        os_log("setInputMode: %{public}@", log: Self.log, type: .debug, mode)
        inputMode = mode
        switch mode {
        case "call":
            astService?.startAudio()
        case "short_voice":
            break
        default:
            astService?.stopAudio()
        }
    }

    public func interrupt() {
        astService?.interrupt()
        transitionTo(.connected)
    }

    public func release() {
        astService?.release()
        astService = nil
    }

    // ── External audio (call translation) ─────────────────────────────────

    public func externalAudioCapability() -> ExternalAudioCapability {
        astService?.externalAudioCapability() ?? .unsupported
    }

    public func startExternalAudio(format: ExternalAudioFormat, sink: ExternalAudioSink) throws {
        guard let svc = astService else {
            throw NativeServiceError.invalidConfig("ast service not initialised")
        }
        try svc.startExternalAudio(format: format, sink: sink)
    }

    public func pushExternalAudioFrame(_ frame: Data) {
        astService?.pushExternalAudioFrame(frame)
    }

    public func stopExternalAudio() {
        astService?.stopExternalAudio()
    }

    // ── Internal ──────────────────────────────────────────────────────────

    fileprivate func transitionTo(_ newState: State) {
        stateLock.lock()
        state = newState
        stateLock.unlock()
        eventSink?.onStateChanged(
            sessionId: config.agentId,
            state: newState.rawValue,
            requestId: nil
        )
    }

    fileprivate func sink() -> AgentEventSink? { eventSink }
    fileprivate func agentId() -> String { config.agentId }
    fileprivate func currentInputMode() -> String { inputMode }
    fileprivate func currentService() -> NativeAstService? { astService }

    fileprivate func setActiveTranslationRequestId(_ id: String?) { activeTranslationRequestId = id }
    fileprivate func getActiveTranslationRequestId() -> String? { activeTranslationRequestId }
    fileprivate func setLastTranslatedText(_ text: String) { lastTranslatedText = text }
    fileprivate func getLastTranslatedText() -> String { lastTranslatedText }
}

// MARK: - AstCallback adapter

private final class AstCallbackAdapter: AstCallback {
    private weak var owner: AstTranslateAgentSession?

    init(owner: AstTranslateAgentSession) { self.owner = owner }

    func onConnected() {
        guard let owner = owner else { return }
        owner.transitionTo(.connected)
        owner.sink()?.onConnectionStateChanged(
            sessionId: owner.agentId(),
            state: "connected",
            errorMessage: nil
        )
        owner.sink()?.onAgentReady(sessionId: owner.agentId(), ready: true)
        if owner.currentInputMode() == "call" {
            owner.currentService()?.startAudio()
        }
    }

    func onDisconnected() {
        guard let owner = owner else { return }
        owner.transitionTo(.idle)
        owner.sink()?.onConnectionStateChanged(
            sessionId: owner.agentId(),
            state: "disconnected",
            errorMessage: nil
        )
    }

    func onRecognitionStart(role: AstRole, requestId: String) {
        // No-op — first partial/firstToken event lazily creates the bubble.
    }

    func onRecognizing(role: AstRole, requestId: String, text: String) {
        guard let owner = owner else { return }
        switch role {
        case .source:
            owner.sink()?.onSttEvent(SttEventData(
                sessionId: owner.agentId(),
                requestId: "",
                kind: "partialResult",
                text: text
            ))
        case .translated:
            owner.setActiveTranslationRequestId(requestId)
            owner.setLastTranslatedText(text)
            owner.sink()?.onLlmEvent(LlmEventData(
                sessionId: owner.agentId(),
                requestId: requestId,
                kind: "firstToken",
                textDelta: text
            ))
        }
    }

    func onRecognized(role: AstRole, requestId: String, text: String) {
        guard let owner = owner else { return }
        switch role {
        case .source:
            owner.sink()?.onSttEvent(SttEventData(
                sessionId: owner.agentId(),
                requestId: requestId,
                kind: "finalResult",
                text: text
            ))
        case .translated:
            owner.setActiveTranslationRequestId(requestId)
            owner.setLastTranslatedText(text)
            owner.sink()?.onLlmEvent(LlmEventData(
                sessionId: owner.agentId(),
                requestId: requestId,
                kind: "firstToken",
                textDelta: text
            ))
        }
    }

    func onRecognitionDone(role: AstRole, requestId: String) {
        // No-op — round-level closure handled by recognitionEnd.
    }

    func onRecognitionEnd(requestId: String) {
        guard let owner = owner else { return }
        let transId = owner.getActiveTranslationRequestId()
        let transText = owner.getLastTranslatedText()
        if let id = transId, !transText.trimmingCharacters(in: .whitespaces).isEmpty {
            owner.sink()?.onLlmEvent(LlmEventData(
                sessionId: owner.agentId(),
                requestId: id,
                kind: "done",
                fullText: transText
            ))
        }
        owner.setActiveTranslationRequestId(nil)
        owner.setLastTranslatedText("")
    }

    func onRecognitionError(requestId: String?, role: AstRole?, code: String, message: String) {
        guard let owner = owner else { return }
        owner.sink()?.onError(
            sessionId: owner.agentId(),
            errorCode: code,
            message: message,
            requestId: requestId
        )
    }

    func onError(code: String, message: String) {
        guard let owner = owner else { return }
        owner.transitionTo(.error)
        owner.sink()?.onConnectionStateChanged(
            sessionId: owner.agentId(),
            state: "error",
            errorMessage: message
        )
        owner.sink()?.onError(
            sessionId: owner.agentId(),
            errorCode: code,
            message: message,
            requestId: nil
        )
        owner.sink()?.onAgentReady(
            sessionId: owner.agentId(),
            ready: false,
            errorCode: code,
            errorMessage: message
        )
    }
}
